packages/components/src/number-control/test/index.tsx

```diff
@@ -0,0 +1,616 @@
+			await waitFor( () => {
+				expect( onChangeSpy ).toHaveBeenCalledTimes( 3 );
+			} );
+
+			// First call: clear the input
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 1, '', true );
+			// Second call: type '1'
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 2, '1', false );
+			// Third call: clamp value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 3, 4, true );
+		} );
+
+		it( 'should call onChange callback when value is not valid', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+			const onChangeSpy = jest.fn();
+
+			render(
+				<NumberControl
+					value={ 5 }
+					min={ 1 }
+					max={ 10 }
+					onChange={ ( v, extra ) =>
+						onChangeSpy(
+							v,
+							( extra.event.target as HTMLInputElement ).validity
+								.valid
+						)
+					}
+				/>
+			);
+
+			const input = getInput();
+
+			await user.clear( input );
+			await user.type( input, '14' );
+
+			expect( input ).toHaveValue( 14 );
+
+			await user.keyboard( '[Enter]' );
+
+			expect( input ).toHaveValue( 10 );
+
+			expect( onChangeSpy ).toHaveBeenCalledTimes( 4 );
+
+			// First call: clear value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 1, '', true );
+			// Second call: invalid, unclamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 2, '1', true );
+			// Third call: invalid, unclamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 3, '14', false );
+			// Fourth call: valid, clamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 4, 10, true );
+		} );
+	} );
+
+	describe( 'Validation', () => {
+		it( 'should clamp value within range on ENTER keypress', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+
+			await user.clear( input );
+			await user.type( input, '-100' );
+			await user.keyboard( '[Enter]' );
+
+			/**
+			 * This is zero because the value has been adjusted to
+			 * respect the min/max range of the input.
+			 */
+			expect( input ).toHaveValue( 0 );
+		} );
+
+		it( 'should clamp value within range on blur', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, '41' );
+
+			// Before blurring, the value is still un-clamped
+			expect( input ).toHaveValue( 41 );
+
+			// Blur the input
+			await user.click( document.body );
+
+			// After blur, value is clamped
+			expect( input ).toHaveValue( 10 );
+		} );
+
+		it( 'should parse to number value on ENTER keypress when required', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ true } /> );
```

Comment:
We could remove some redundant `={ true }` around the codebase.

```suggestion
			render( <NumberControl value={ 5 } required /> );
```

Reply:
Yup, made those changes too